metaflow/extension_support.py

```diff
@@ -0,0 +1,441 @@
+    )
+    return lazy_load_custom_modules
+
+
+def lazy_load_aliases(aliases):
+    if aliases:
+        sys.meta_path = [_LazyLoader(aliases)] + sys.meta_path
+
+
+def multiload_globals(modules, dst_globals):
+    for m in modules:
+        load_globals(m.module, dst_globals)
+
+
+def multiload_all(modules, extension_point, dst_globals):
+    lazy_load_custom_modules = {}
+    for m in modules:
+        lazy_load_custom_modules.update(
+            alias_submodules(m.module, m.tl_package, extension_point)
+        )
+        load_globals(m.module, dst_globals)
+    lazy_load_aliases(lazy_load_custom_modules)
+
+
+_py_ver = sys.version_info[0] * 10 + sys.version_info[1]
+_mfext_supported = False
+
+if _py_ver >= 34:
+    import importlib.util
+    from importlib.machinery import ModuleSpec
+
+    if _py_ver >= 38:
+        from importlib import metadata
+    else:
+        import importlib_metadata as metadata
+    _mfext_supported = True
+else:
+    # Something random so there is no syntax error
+    ModuleSpec = None
+
+# IMPORTANT: More specific paths must appear FIRST (before any less specific one)
+_extension_points = [
+    "plugins.env_escape",
+    "plugins.cards",
+    "config",
+    "datatools",
+    "exceptions",
+    "plugins",
+    "toplevel",
+]
+
+
+def _ext_debug(*args, **kwargs):
+    if METAFLOW_DEBUG_EXT_MECHANISM:
+        print(*args, **kwargs)
+
+
+def _get_extension_packages():
+    if not _mfext_supported:
+        return []
+
+    # If we have an INFO file with the appropriate information (if running from a saved
+    # code package for example), we use that directly
+    # Pre-compute on _extension_points
+    from metaflow import INFO_FILE
+
+    try:
+        with open(INFO_FILE, "r") as contents:
+            all_pkg, ext_to_pkg = json.load(contents).get("ext_info", (None, None))
+            if all_pkg is not None and ext_to_pkg is not None:
+                # We need to properly convert stuff in ext_to_pkg
+                for k, v in ext_to_pkg.items():
+                    v = [MFExtPackage(*d) for d in v]
+                    ext_to_pkg[k] = v
+                return all_pkg, ext_to_pkg
+    except IOError:
+        pass
+
+    _list_ext_points = [x.split(".") for x in _extension_points]
+    # TODO: This relies only on requirements to determine import order; we may want
+    # other ways of specifying "load me after this if it exists" without depending on
+    # the package. One way would be to rely on the description and have that info there.
+    # Not sure of the use though so maybe we can skip for now.
+    mf_ext_packages = []
+    extension_points_to_pkg = defaultdict(dict)
+    for dist in metadata.distributions():
+        if any(
+            [pkg == EXT_PKG for pkg in (dist.read_text("top_level.txt") or "").split()]
+        ):
+            _ext_debug("Found extension package '%s'..." % dist.metadata["Name"])
+            mf_ext_packages.append(dist.metadata["Name"])
```

Comment:
There can be more than one top-level package in `dist`. See the example in the comments left earlier about `_pkgs_per_extension_point`